File: 프로그래머스/2/142085. 디펜스 게임/디펜스 게임.py
import heapq
def solution(n, k, enemy):
    answer = 0
    
    if k >= len(enemy) :
        return len(enemy)
    
    use = []
    
    for i in range(len(enemy)):
        if k>0:
            k-=1
            answer +=1
            heapq.heappush(use, enemy[i])
        else:
            heapq.heappush(use,enemy[i])
            n -= heapq.heappop(use)
            if n<0:
                break
            else:
                answer+=1
    return answer

File: 프로그래머스/2/142085. 디펜스 게임/디펜스 게임.java
import java.util.*;
class Solution {
    public int solution(int n, int k, int[] enemy) {
        int answer = 0;
        
        Queue<Integer> pq= new PriorityQueue<>(Collections.reverseOrder());
        
        for(int i=0; i<enemy.length;i++){
            n-=enemy[i];
            pq.add(enemy[i]);
            
            if(n<0){
                if(k>0){
                    n+=pq.poll();
                    k-=1;
                }else{
                    break;
                }
            }
            answer++;
        }
        
        
        return answer;
    }
}